mantidproject/mantid
# TOF Powder group calibration scripts

**Description of work.**

Workflow for [PD240](https://code.ornl.gov/sns-hfir-scse/diffraction/powder/powder-diffraction/-/issues/240)


**To test:**

The usage example should be a good test to run through.

Here is a real example, just loading a few banks and creating groups by bank

```python
from mantid.simpleapi import *
from Calibration.tofpd import group_calibration

data_file = 'NOM_144974'

data = LoadEventNexus(data_file, BankName='bank45,bank46,bank47,bank48,bank49', NumberOfBins=1)
data = MaskDetectorsIf(data, Operator='Less', Value=25000)

groups, _, _ = CreateGroupingWorkspace(data, GroupDetectorsBy='bank', OutputWorkspace='group_banks')

output_workspace_basename = 'calibration'

cc_diffcal = group_calibration.cc_calibrate_groups(data,
                                                   groups,
                                                   output_workspace_basename)

# Check CC calibration
data_d = ConvertUnits(data, Target='dSpacing')
data_d = Rebin(data_d, '0.25,0.002,2.25')
data_aligned = AlignDetectors(data, CalibrationWorkspace=cc_diffcal)
data_aligned = Rebin(data_aligned, '0.25,0.002,2.25')

diffcal, mask = group_calibration.pdcalibration_groups(data,
                                                       groups,
                                                       cc_diffcal,
                                                       output_workspace_basename,
                                                       PeakWindow=0.4)

# Check final calibration
data_aligned2 = AlignDetectors(data, CalibrationWorkspace=diffcal)
data_aligned2 = Rebin(data_aligned2, '0.25,0.002,2.25')
```

You can look at the starting `data_d` workspace and compare to after CC `data_aligned` and after CC+PD `data_aligned2`. The Colorfill plot is good to look at.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
